relay-server/src/actors/processor.rs

```diff
@@ -1182,17 +1197,13 @@ impl EnvelopeProcessorService {
             }
             _ => ItemAction::Keep,
         });
-        if let Some(event) = state.event.value_mut() {
-            if event.ty.value() == Some(&EventType::Transaction) {
-                let contexts = event.contexts.get_or_insert_with(Contexts::new);
-                // If we found a profile, add its ID to the profile context on the transaction.
-                if let Some(profile_id) = found_profile_id {
-                    contexts.add(ProfileContext {
-                        profile_id: Annotated::new(profile_id),
-                    });
-                // If not, we delete the profile context.
-                } else {
-                    contexts.remove::<ProfileContext>();
+        if found_profile_id.is_none() {
+            // Remove profile context from event.
+            if let Some(event) = state.event.value_mut() {
+                if event.ty.value() == Some(&EventType::Transaction) {
+                    if let Some(contexts) = event.contexts.value_mut() {
+                        contexts.remove::<ProfileContext>();
+                    }
```

Comment:
When profile processing fails, the profile ID is removed here. This means that some (invalid) profiles that were opted in by a dynamic sampling rule based on `contexts.profile.profile_id` might end up _without_ a profile ID in the final payload.

Reply:
So this implies that we are performing sampling on profile ids that we don't know whether they are valid of not, am I right?

From the code flow, I see that we do filter the profile with context injection, we then move onto sampling (assuming the profile id refers to a valid profile), and then we process the profile and remove the context in case it's invalid?

Is this an expected behavior? Just curious

Reply:
> So this implies that we are performing sampling on profile ids that we don't know whether they are valid of not, am I right?

That is correct and a bit unfortunate. To fix this we would have to move the entirety of `process_profiles` to non-processing relays (and before dynamic sampling), and I'm not sure if that's currently possible. cc @jan-auer

Reply:
Sounds good, thanks for the explanation!